share/rocm/cmake/ROCMInstallTargets.cmake

```diff
@@ -4,11 +4,14 @@
 
 cmake_policy(SET CMP0057 NEW)
 
+# Default libdir to "lib", this skips GNUInstallDirs from trying to take a guess if it's unset:
+set(CMAKE_INSTALL_LIBDIR "lib" CACHE STRING "Library install directory")
+
 include(CMakeParseArguments)
 include(GNUInstallDirs)
 include(ROCMPackageConfigHelpers)
 
-set(ROCM_INSTALL_LIBDIR lib)
+set(ROCM_INSTALL_LIBDIR ${CMAKE_INSTALL_LIBDIR})
```

Comment:
Why are we using cached variable for this?

Reply:
I think it would be better to have a variable to enable non-std layout perhaps we could have `ROCM_USE_NON_STD_LAYOUT` and devops could set that instead. I dont think `CMAKE_INSTALL_LIBDIR` was meant to be used as a cached variable, so I am not sure if it might have unintended consequences. 

I would rather have the correct layout by default and then rocm can enable the non-standard layout for its packaging. We shouldn't require users to set either `ROCM_USE_STD_LAYOUT` or `CMAKE_INSTALL_LIBDIR` to get the correct layout.

Reply:
So CMAKE_INSTALL_LIBDIR was made to be overriden, since some distros have creative libdir paths, see debian ;). The idea is that GNUInstallDirs tries to set some sensible defaults, but doesn't change the value if set. I'm not sure if the cached variable is strictly required though, I did it more just to be safe as I'm not the best with cmake.

The implementation in GNUInstallDirs is (i trimmed out some lines):

if(__system_type_for_install STREQUAL "debian")
...
  set(_LIBDIR_DEFAULT "lib/${CMAKE_LIBRARY_ARCHITECTURE}")
...
elseif(NOT DEFINED __system_type_for_install)
...
  set(_LIBDIR_DEFAULT "lib64")
...
endif()
...
if(NOT DEFINED CMAKE_INSTALL_LIBDIR)
  set(CMAKE_INSTALL_LIBDIR "${_LIBDIR_DEFAULT}" CACHE PATH "Object code libraries (${_LIBDIR_DEFAULT})")
...
endif()

Reply:
That's an interesting perspective. @Mystro256 and I were looking at this more narrowly. I certainly like the idea of the ROCm libraries following a more normal build/install process by default and just having an option to switch to the `/opt/rocm` alternative layout.

That would be a pretty significant change, though. We'd probably need to start by defaulting the option to the `/opt/rocm` layout. Then we would update our build scripts to explicitly specify that option. Finally, once everyone had specified it, we could change the default.

> Why are we using cached variable for this?

CMake itself defines CMAKE_INSTALL_LIBDIR as a cache variable when you include GNUInstallDirs. https://github.com/Kitware/CMake/blob/v3.23.1/Modules/GNUInstallDirs.cmake#L297

The reasoning is basically so that distros can pass `-DCMAKE_INSTALL_LIBDIR=<whatever>` and set it to the value that their style guide requires without having to patch our build code.

> I would rather have the correct layout by default and then rocm can enable the non-standard layout for its packaging.

I agree, but that is a big change. I don't think Jeremy's suggestion conflicts with moving to the model you suggest.

Reply:
Personally if you ask me, I'd prefer just using GNUInstallDirs as-is, but have ROCm-CI (or whoever) override the values of CMAKE_INSTALL_PREFIX=/opt/rocm and CMAKE_INSTALL_LIBDIR=lib... but packaging didn't seem to like the idea, instead Freddy proposed the solution of just using a cached value of "lib" like I did in my patch. I think they want the default install case to fit the ROCm design to avoid confusion over customers compiling from source who don't understand that LIBDIR could be lib64 or lib or lib/somevalue.

At the end of the day though, the distros don't care as long as the paths aren't hardcoded, so those "FORCE" lines are the problem I was attempting to remedy.

Reply:
Alright, cache var is fine, we should just probably remove the `ROCM_INSTALL_LIBDIR` variable since its not needed.